File: dref-core/src/main/scala/io/github/tobia80/dref/LockValue.scala
package io.github.tobia80.dref

import java.nio.ByteBuffer

/** Wire format for distributed lock tokens. Both Scala and Rust clients
  * store lock ownership as an 8-byte big-endian `Long` so mixed-language
  * clusters can detect stolen locks consistently.
  */
object LockValue {

  def toBytes(value: Long): Array[Byte] = {
    val buf = ByteBuffer.allocate(8)
    buf.putLong(value)
    buf.array()
  }

  def fromBytes(bytes: Array[Byte]): Long =
    ByteBuffer.wrap(bytes).getLong()
}

File: rust/dref-core/src/lock_value.rs
//! Wire format for distributed lock tokens. Both Scala and Rust clients store
//! lock ownership as an 8-byte big-endian i64 so mixed-language clusters can
//! detect stolen locks consistently.

/// Encode a lock token to its on-the-wire representation.
pub fn to_bytes(value: i64) -> [u8; 8] {
    value.to_be_bytes()
}

/// Decode a lock token from its on-the-wire representation.
pub fn from_bytes(bytes: &[u8]) -> Option<i64> {
    let arr: [u8; 8] = bytes.try_into().ok()?;
    Some(i64::from_be_bytes(arr))
}
